### PR TITLE
feat(labels): Adds .metric and .label syntax

### PR DESCRIPTION
**Pull request by pselle**
- Adds .metric and .label syntax and docs to this agent
- Update docs for custom metrics
- Adds docs for labels

### CODE REVIEW

File: iopipe/context.py [hunk 1]
```diff
@@ -34,6 +34,14 @@ def log(self, key, value):
 
         self.instance.report.custom_metrics.append(event)
 
+    def log(self, key, value):
+        warnings.warn('context.iopipe.log is deprecated and will be removed in a future version, '
+                      'please use context.iopipe.metric')
+        self.metric(key, value)
+
+    def tag(self, key):
```
Comment (kolanos) [inline iopipe/context.py]: How about `def tag(self, *keys):`?

File: tests/test_agent.py [hunk 2]
```diff
@@ -40,7 +40,20 @@ def test_custom_metrics(mock_send_report, handler_with_events, mock_context):
     iopipe, handler = handler_with_events
     handler(None, mock_context)
 
+    assert len(iopipe.report.custom_metrics) == 7
+    # Decimals are converted to strings
+    assert iopipe.report.custom_metrics[6]['s'] == '12.300000000000000710542735760100185871124267578125'
```
Comment (kolanos) [inline tests/test_agent.py]: Does the node agent cast to floats? Maybe we should make this behavior consistent.

File: iopipe/context.py [hunk 3]
```diff
@@ -34,6 +42,23 @@ def log(self, key, value):
 
         self.instance.report.custom_metrics.append(event)
 
+    def log(self, key, value):
```
Comment (kolanos) [inline iopipe/context.py]: This could be simplified to just `log = metric`.

Comment (kolanos) [inline iopipe/context.py]: This could be simplified to just `log = metric`.

File: iopipe/context.py [hunk 4]
```diff
@@ -34,6 +42,23 @@ def log(self, key, value):
...
+        self.metric(key, value)
+
+    def label(self, name):
+        if not isinstance(name, str):
```
Comment (kolanos) [inline iopipe/context.py]: This check will be problematic as in Python 2.7 `unicode` is also a string type, but does not exist in Python 3.x. You can borrow the string types handling in the popular `six` library to make this Python 2/3 compatible.

File: iopipe/report.py [hunk 5]
```diff
@@ -37,6 +37,7 @@ def __init__(self, config, context):
         self.config = config
         self.context = context
         self.custom_metrics = []
+        self.labels = []
```
Comment (kolanos) [inline iopipe/report.py]: Shouldn't the `[]` be a `set()`?

File: iopipe/context.py [hunk 6]
```diff
@@ -34,6 +42,23 @@ def log(self, key, value):
...
+                          'This label will not be recorded.')
+            return
+
+        if len(name) > constants.METRIC_NAME_LIMIT:
+            warnings.warn('Label of name %s is longer than allowed limit of '
+                          '%s characters. '
+                          'This label will not be recorded.' % (name, constants.METRIC_NAME_LIMIT))
+            return
+
+        self.instance.report.labels.append(name)
```
Comment (kolanos) [inline iopipe/context.py]: If this is a `set()` then `append` should be `add`.

File: iopipe/report.py (hunk 5 above)
Comment (pselle) [reply iopipe/report.py]: Yes! Will do and add tests appropriately

File: iopipe/context.py (hunk 4 above)
Comment (pselle) [reply iopipe/context.py]: should I do `type(name) == 'str'`?  that means we wouldn't have to add the six library to deps

**Comment by pselle** [comment]
@kolanos and thanks for catching the unicode! I added a test for it as well

**Pull request merged**